docs: proposal for providing PV key rotation

black-dragon74 opened the pull request:
The design and implementation details for rotating the encryption keys for volumes.

black-dragon74 committed:
doc: proposal for providing PV key rotation

The design and implementation details for rotating
the encryption keys for volumes.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

Rakshith-R commented on docs/design/proposals/pv-key-rotation.md:
```diff
@@ -0,0 +1,55 @@
+- Create a `rbdvolume` object using volume ID, this is done by `GenVolFromVolID`.
+- Fetch the current key from the KMS, it is needed for subsequent LUKS operations.
+- Get the device path for the volume by calling `waitForPath` as all LUKS operations require the device path.
+- Add the fetched key to LUKS slot 1, this will serve as a backup of the key.
+- Generate a new key and store it locally. It will be updated in the KMS at later steps.
+- Remove the exsitng key from slot 0 upon verifying that the key in KMS == the key in slot 0.
+- Add new key to slot 0 and then call `LuksVerifyKey` to verify that the slot was successfully updated.
+- Update the new key in the KMS.
+- Fetch the key again and verify that the key in KMS == the new key we generated.
+- We can now remove the backup key from slot 1.
```
Let's keep the process the same as the one implemented here https://github.com/rook/rook/blob/ff448e11d8836388e976d0b7de74f517cd76e275/pkg/daemon/ceph/osd/key_rotation.go#L31

black-dragon74 replied:
You mean the wordings of this doc? As the steps are essentially the same.

Rakshith-R replied:
LuksVerify function is not used in the above linked function.
Let's keep it the same so as to avoid extra work and having different steps in each place.

black-dragon74 replied:
Ack.

Rakshith-R commented on docs/design/proposals/pv-key-rotation.md:
```diff
@@ -0,0 +1,55 @@
+
+- Create a `rbdvolume` object using volume ID, this is done by `GenVolFromVolID`.
+- Fetch the current key from the KMS, it is needed for subsequent LUKS operations.
+- Get the device path for the volume by calling `waitForPath` as all LUKS operations require the device path.
```
How are you going to determine the device path ?

black-dragon74 replied:
As per my current understanding, calling `waitForPath` defined in `rbd_attach.go` can get us the device path. Am I wrong to assume that?

Rakshith-R commented on docs/design/proposals/pv-key-rotation.md:
```diff
@@ -0,0 +1,55 @@
+
+The `VolumeEncryption` struct is modified to make `generateNewEncryptionPassphrase` a public member function.
+
+A metadata is set on the RBD image to indicate that the image is being processed for keyrotation. Presence of this metadata will prevent the same image being processed again.
```
What's the purpose of setting metadata and which server controller or node will set this / respond to this ? 
When will the rotation be triggered?

black-dragon74 replied:
We will register this service only on the node server hence the node server will set the metadata. Before doing that it will ensure that it is not already present.

The purpose of setting this metadata is to serve as a indication to avoid processing of the same PVC while it is already being processed. Might be extraneous though. If so, we can ditch this step.

The rotation will be triggered when the node server receives the RPC call from csi-addons. Which will be in turn be triggered by the cron schedule defined in the annotated storage class.

Rakshith-R replied:
The volume/operation locks which we have will serve that purpose ? 

I guess metadata is required if we have rbd mounted in multiple nodes but even with metadata set if controller pings another nodeplugin after first request then the status will get updated to complete?

black-dragon74 replied:
In case of RWX volumes we will send the request to only one of the nodes, as the underlying device is the same and doing it on one node will do it for all.

Ditching the metadata part then :)

Rakshith-R replied:
> In case of RWX volumes we will send the request to only one of the nodes, as the underlying device is the same and doing it on one node will do it for all.
> 
> Ditching the metadata part then :)

Yea, one more problem will be how to decide which node to target and send requests only to that.

black-dragon74 replied:
What I have so far is to get a list of vol attachments and see if the node advertises KeyRotate identity. If so, we send the request to it. We can safely ignore the rest of the nodes in the attachment list. This will work for both type of volumes as we need to find out the node to send rpc requests to anyways. WDYT?

Rakshith-R replied:
My main concern is request timing out and then we may choose another node in a follow up request. Operation locks cannot help in this case.

Madhu-1 replied:
what if we have a a recovery key generated by the controller and store it in the KMS temporarily until the existing key is updated in KMS? (we still need to check and see if this is a feasible and accepted solution)

black-dragon74 replied:
Just a thought, what about we requeue the reconcile and the next time we do not forward the request to the node that timed out? Leaning towards this as we won't have to change the existing KMS impl this way and we'll have much less processing on csi driver side of things.

Madhu-1 replied:
You need to think about the side effect it going to have multiple key or key not matching which is in KMS with the one actual in-use.

Rakshith-R commented on docs/design/proposals/pv-key-rotation.md:
```diff
@@ -0,0 +1,55 @@
+
+### Implementation Details
+
+The encryption key rotation request will contain with it the volume ID, credentials and secrets.
```
credentials and secrets are one and the same here ?

black-dragon74 replied:
I will update this. We use the secrets to generate the credentials. The request will contain volumeID and secrets.

Madhu-1 commented on docs/design/proposals/pv-key-rotation.md:
```diff
@@ -0,0 +1,55 @@
+
+The following steps are followed to process the device for key rotation:
+
+- Create a `rbdvolume` object using volume ID, this is done by `GenVolFromVolID`.
```
please separate the controller and node-plugin process here and also mention how you will handle RWX PVC as well.

black-dragon74 replied:
Ack.

Madhu-1 commented on docs/design/proposals/pv-key-rotation.md:
```diff
@@ -0,0 +1,55 @@
+# Encryption Key Rotation
```
update the file name `rbd-pv-key-rotation` as this is specific to rbd.

black-dragon74 replied:
Ack.

black-dragon74 committed:
doc: Update docs for rbd-pv-key-rotation

This commit updates the key rotation docs with the following changes:

- Do not call LuksVerify
- Mention specifics of RWX volumes
- Rename the file to represent RBD backed volumes

Signed-off-by: black-dragon74 <niryadav@redhat.com>

Madhu-1 commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,57 @@
+
+The secrets are used to generate the credentials for autheticating aginst a ceph cluster.
+
+These values are then used to call `GenVolFromVolID` to get the rbdVolume structure.
```
`aginst` to `against`

Madhu-1 commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,57 @@
+
+These order of the above steps guarantees that we always have one key that can unlock the encrypted volume.
+
+In case of RWX volumes, we send the request to ONLY one of the nodes where the volume is mounted. As the underlying device is same, rotating the key on any one of the nodes will suffice and prevent us from additional performance overheads.
```
This is not yet decided right? Please mark RWX as out of scope/future from this one.

Madhu-1 commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,57 @@
+
+The proposed solution in this document, is to address the periodic rotation of encryption keys for encrypted volumes.
+
+This document outlines the rotation steps for PVCs backed by RBD and will be updated with other volume types as they are supported.
```
`will be updated with other volume types as they are supported` what is other volume types here?

black-dragon74 replied:
Cephfs backed volumes encrypted using fscrypt.

Madhu-1 replied:
This doc is for rbd only as the doc name suggest, we don't need to cover cephfs specific things in this doc

Madhu-1 commented:
@black-dragon74 please fix the CI failures

nixpanic commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,77 @@
+encryption keys (KEKs) for encrypted volumes in Ceph-CSI.
+
+Support for rotating keys on RWX/ROX volumes is out of scope for now
+and shall be added later.
```
Also add that filesystem encryption with `fscrypt` is out of scope, this is only for block-device encryption with LUKS. The RBD backend supports both, CephFS only supports `fscrypt`, NFS supports none.

nixpanic commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,77 @@
+
+## Proposed Solution
+
+The proposed solution in this document, is to address the periodic rotation
```
How is this a proposal for `periodic` rotation?  There is no scheduler that does the key rotation at certain intervals/periods. This is a description of what the CSI-Addons API requests will do, who/whatever the API calls at what intervals/periods is out of scope for the implementation of the API.

nixpanic commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,77 @@
+
+This feature builds upon the foundation laid by encrypted pvcs.
+
+An existing storage class can be annotated with
```
this is not part of Ceph-CSI, you can probably refer to the documentation in the Kubernetes CSI-Addons repository for these details.

nixpanic commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,77 @@
+- Add the fetched key to LUKS slot 1, this will serve as a backup of the key.
+- Generate a new key and store it locally. It will be updated
+  in the KMS at later steps.
+- Remove the exsitng key from slot 0 upon verifying that the
```
typo: exsitng

nixpanic commented on docs/design/proposals/rbd-pv-key-rotation.md:
```diff
@@ -0,0 +1,77 @@
+  key in KMS == the new key we generated.
+- We can now remove the backup key from slot 1.
+
+These order of the above steps guarantees that we always have one key
```
note that the key in the KMS can always be used to unlock the volume, a small but important detail :wink:

nixpanic commented:
@mergifyio queue
